meta-lmp-base/recipes-containers/docker/files/0001-tarexport-Load-image-without-layer-copying.patch

```diff
@@ -0,0 +1,106 @@
++			layerPath, err := safePathWithTmp(m.LayerRootDir, tmpDir, m.Layers[i])
+ 			if err != nil {
+ 				return err
+ 			}
+@@ -123,14 +123,21 @@ func (l *tarexporter) Load(inTar io.ReadCloser, outStream io.Writer, quiet bool)
+ 		for _, repoTag := range m.RepoTags {
+ 			named, err := reference.ParseNormalizedNamed(repoTag)
+ 			if err != nil {
+-				return err
++				return fmt.Errorf("Invalid image reference: %s", repoTag)
+ 			}
+-			ref, ok := named.(reference.NamedTagged)
+-			if !ok {
+-				return fmt.Errorf("invalid tag %q", repoTag)
++			switch ref := named.(type) {
++				case reference.NamedTagged: {
++					l.setLoadedTag(ref, imgID.Digest(), outStream)
++					outStream.Write([]byte(fmt.Sprintf("Loaded image with tag: %s\n", reference.FamiliarString(ref))))
++				}
++				case reference.Canonical: {
++					l.setLoadedDigest(ref, imgID.Digest(), outStream)
++					outStream.Write([]byte(fmt.Sprintf("Loaded image with digest: %s\n", reference.FamiliarString(ref))))
++				}
++				default: {
++					outStream.Write([]byte(fmt.Sprintf("Unsupported image reference type: %s\n", repoTag)))
++				}
+ 			}
+-			l.setLoadedTag(ref, imgID.Digest(), outStream)
+-			outStream.Write([]byte(fmt.Sprintf("Loaded image: %s\n", reference.FamiliarString(ref))))
+ 			imageRefCount++
+ 		}
+ 
+@@ -211,6 +218,14 @@ func (l *tarexporter) setLoadedTag(ref reference.Named, imgID digest.Digest, out
+ 	return l.rs.AddTag(ref, imgID, true)
+ }
+ 
++func (l *tarexporter) setLoadedDigest(ref reference.Canonical, imgID digest.Digest, outStream io.Writer) error {
++	if prevID, err := l.rs.Get(ref); err == nil && prevID != imgID {
++		fmt.Fprintf(outStream, "The image %s already exists, renaming the old one with ID %s to empty string\n", reference.FamiliarString(ref), string(prevID)) // todo: this message is wrong in case of multiple tags
++	}
++
++	return l.rs.AddDigest(ref, imgID, true)
++}
++
+ func (l *tarexporter) legacyLoad(tmpDir string, outStream io.Writer, progressOutput progress.Output) error {
+ 	if runtime.GOOS == "windows" {
+ 		return errors.New("Windows does not support legacy loading of images")
+@@ -371,6 +386,14 @@ func safePath(base, path string) (string, error) {
+ 	return symlink.FollowSymlinkInScope(filepath.Join(base, path), base)
+ }
+ 
++func safePathWithTmp(baseDir, tmp, path string) (string, error) {
++	base := baseDir
++	if len(base) == 0 {
++		base = tmp
++	}
++	return symlink.FollowSymlinkInScope(filepath.Join(base, path), base)
++}
++
+ type parentLink struct {
+ 	id, parentID image.ID
+ }
+diff --git a/image/tarexport/tarexport.go b/image/tarexport/tarexport.go
+index 5bcad2265c..f53a1c4be6 100644
+--- a/image/tarexport/tarexport.go
++++ b/image/tarexport/tarexport.go
+@@ -21,6 +21,7 @@ type manifestItem struct {
+ 	Layers       []string
+ 	Parent       image.ID                                 `json:",omitempty"`
+ 	LayerSources map[layer.DiffID]distribution.Descriptor `json:",omitempty"`
++	LayerRootDir string `json:"layer_root_dir,omitempty"`
```

Comment:
@doanac Just heads up. By extending the manifest passed in the TAR archive/stream to `/image/load` (`docker load` command) we kind of instruct the docker's tar exporter to load layer blobs from a specific local directory. As result, no need to pass layer blobs through the TAR stream in the `HTTP POST /image/load` body.